File: src/event.rs
use serde::{Serialize, ser::SerializeStruct};

use crate::model::{Address, Aircraft};

#[derive(Clone)]
pub enum AircraftEvent {
    Present(Aircraft),
    Entered(Aircraft),
    Left(Aircraft),
    Landed(Aircraft),
    TookOff(Aircraft),
    ChangedAddress(Aircraft, Address),
}

impl AircraftEvent {
    fn action(&self) -> &str {
        match self {
            Self::Entered(_) => "entered",
            Self::Left(_) => "left",
            Self::Landed(_) => "landed",
            Self::TookOff(_) => "took_off",
            Self::Present(_) => "present",
            Self::ChangedAddress(_, _) => "changed_address",
        }
    }
}

impl Serialize for AircraftEvent {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        // Important: increase the state len if new fields are added
        match self {
            Self::Present(ac)
            | Self::Entered(ac)
            | Self::Left(ac)
            | Self::Landed(ac)
            | Self::TookOff(ac) => {
                let mut state = serializer.serialize_struct("AircraftEvent", 10)?;
                state.serialize_field("action", self.action())?;
                state.serialize_field("icao24", &ac.icao24)?;
                state.serialize_field("callsign", ac.callsign.as_deref().unwrap_or_default())?;
                state.serialize_field("latitude", &ac.position.lat)?;
                state.serialize_field("longitude", &ac.position.lon)?;
                state.serialize_field("category", &ac.category)?;
                state.serialize_field("on_ground", &ac.on_ground)?;

                state.serialize_field("city", &ac.address.city)?;
                state.serialize_field("state", &ac.address.state)?;
                state.serialize_field("country", &ac.address.country)?;

                state.end()
            }
            Self::ChangedAddress(ac, prev_address) => {
                let mut state = serializer.serialize_struct("AircraftEvent", 13)?;
                state.serialize_field("action", self.action())?;
                state.serialize_field("icao24", &ac.icao24)?;
                state.serialize_field("callsign", ac.callsign.as_deref().unwrap_or_default())?;
                state.serialize_field("latitude", &ac.position.lat)?;
                state.serialize_field("longitude", &ac.position.lon)?;
                state.serialize_field("category", &ac.category)?;
                state.serialize_field("on_ground", &ac.on_ground)?;

                state.serialize_field("city", &ac.address.city)?;
                state.serialize_field("state", &ac.address.state)?;
                state.serialize_field("country", &ac.address.country)?;

                state.serialize_field("prev_city", &prev_address.city)?;
                state.serialize_field("prev_state", &prev_address.state)?;
                state.serialize_field("prev_country", &prev_address.country)?;

                state.end()
            }
        }
    }
}

File: frontend/src/types/event.ts
// Types mirror the backend Serialize impl in airspace-pulse/src/event.rs
// and model.rs. Do not invent fields.

export type AircraftAction =
  | "present"
  | "entered"
  | "left"
  | "landed"
  | "took_off"
  | "changed_address";

export type AircraftCategory =
  | "noInfo"
  | "light"
  | "small"
  | "large"
  | "heavy"
  | "rotorcraft"
  | "glider"
  | "lighterThanAir"
  | "unmannedAerialVehicle"
  | "space"
  | "emergency"
  | "service"
  // backend serializes Unknown(n) as `unknown{n}`; kept permissive
  | `unknown${number}`
  | string;

export interface AircraftEvent {
  action: AircraftAction;
  icao24: string;
  callsign: string;
  latitude: number;
  longitude: number;
  category: AircraftCategory;
  on_ground: boolean;
  // Backend reverses geocode (lat, lon) via a local geocoder. The underlying
  // dataset strips diacritics, so city/state/country values arrive unaccented
  // (e.g. "Sao Paulo" instead of "São Paulo"). Display as-is.
  city: string;
  state: string;
  country: string;
  // Present only on `changed_address` events — the previous address before the
  // aircraft crossed into a new reverse-geocoded cell. Absent on every other
  // action.
  prev_city?: string;
  prev_state?: string;
  prev_country?: string;
}

export type SseStatus = "idle" | "connecting" | "connected" | "disconnected";

// Lifecycle transitions tracked client-side for status derivation.
// Present is a discovery event, not a transition — excluded from this union.
export type TransitionAction = Exclude<AircraftAction, "present">;

export type AircraftStatus =
  | "em-voo"
  | "pousado"
  | "decolando"
  | "recem-pousado"
  | "recem-decolado";

export interface ActiveAircraft {
  // event is the raw backend envelope plus a client-stamped _receivedAt.
  event: AircraftEvent & { _receivedAt: number };
  lastTransitionAt: number | null;
  lastTransitionAction: TransitionAction | null;
}

export interface AirspaceMetrics {
  active: number;
  takeoffs: number;
  landings: number;
  recentEvents: number;
}
